api/app/Modules/Library/Projectors/DraftLyricsProjector.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Modules\Library\Projectors;
+
+use App\Modules\Library\Events\Drafts\Lyrics\DraftLyricsChanged;
+use App\Modules\Library\Events\Drafts\Lyrics\DraftLyricsCreated;
+use App\Modules\Library\Events\Drafts\Lyrics\DraftLyricsDeleted;
+use App\Modules\Library\Events\Drafts\Lyrics\DraftLyricsPublished;
+use App\Modules\Library\Events\Tracks\TrackLyricsChanged;
+use App\Modules\Library\Models\DraftLyrics;
+use Spatie\EventSourcing\EventHandlers\Projectors\Projector;
+
+class DraftLyricsProjector extends Projector
+{
+    public function onDraftLyricsCreated(DraftLyricsCreated $event)
+    {
+        $draftLyrics = new DraftLyrics();
+        $draftLyrics->id = $event->id;
+        $draftLyrics->track_id = $event->trackId;
+        $draftLyrics->format = $event->document->getFormat()->value;
+        $draftLyrics->content = $event->document->getContent();
+        $draftLyrics->save();
+    }
+
+    public function onDraftLyricsChanged(DraftLyricsChanged $event)
+    {
+        $draftLyrics = DraftLyrics::retrieve($event->id);
+        $draftLyrics->format = $event->document->getFormat()->value;
+        $draftLyrics->content = $event->document->getContent();
+        $draftLyrics->save();
+    }
+
+    public function onDraftLyricsPublished(DraftLyricsPublished $event)
+    {
+        $draftLyrics = DraftLyrics::retrieve($event->id);
+        event(new TrackLyricsChanged($draftLyrics->track_id, $event->document));
+        event(new DraftLyricsDeleted($draftLyrics->id));
```

Comment:
I.. I'm not sure this is the right way to go. I think this needs to happen in the controller.

Reply:
Moved this to the model method `publishLyrics`

Reply:
I added deleting of the data within the the projector method itself

New code is below
I did it like this as I felt having draft lyrics deleted was incorrect, as technically the user did not delete, but the system deleted due to publishing. So when publishing, we will only get one event which will be published
```
public function onDraftLyricsPublished(DraftLyricsPublished $event)
    {
        $draftLyrics = DraftLyrics::retrieve($event->id);
        event(new TrackLyricsChanged($draftLyrics->track_id, $event->document));
        $draftLyrics->delete();
    }
```